sale_purchase_requisition/models/sale_order.py

```diff
@@ -0,0 +1,174 @@
+                for pr_line in sale_line.purchase_requisition_line_ids:
+                    if pr_line.requisition_id.state in PURCH_REQ_FROZEN_STATES:
+                        continue  # Purchase Requisition line cannot be modified
+                    pr_line.sudo().write(
+                        {
+                            "product_id": sale_line.product_id.id,
+                            "product_qty": sale_line.product_uom_qty,
+                            "product_uom_id": sale_line.product_uom.id,
+                        }
+                    )
+                # Step 2. Create new purchase requisition line if does not exist
+                if not sale_line.purchase_requisition_line_ids:
+                    editable_purch_reqs = sale.purchase_requisition_ids.filtered(
+                        lambda pr: pr.state not in PURCH_REQ_FROZEN_STATES
+                    )
+                    for purchase_requisition in editable_purch_reqs:
+                        prl_model.sudo().create(
+                            {
+                                "requisition_id": purchase_requisition.id,
+                                "line_order_id": sale_line.id,
+                                "product_id": sale_line.product_id.id,
+                                "product_qty": sale_line.product_uom_qty,
+                                "product_uom_id": sale_line.product_uom.id,
+                            }
+                        )
+        return res
+
+    def action_cancel(self):
+        """Cancel the sale order and cancel or close the purchase requisition."""
+        res = super().action_cancel()
+        purchase_requisitions = self.mapped("purchase_requisition_ids")
+        purchase_requisitions.filtered(
+            lambda pr: pr.state not in {"open", "done", "cancel"}
+        ).action_cancel()
+        purchase_requisitions.filtered(lambda pr: pr.state == "open").action_done()
+        return res
+
+    def _compute_purchase_requisition_sync_warning(self):
+        """Show warning message when order lines are not synced with purchase requisitions."""
+        self.update({"purchase_requisition_sync_warning": None})
+        for sale in self:
+            not_synced_prs = self.env["purchase.requisition"].browse()
+            for line in sale.order_line.filtered("product_id.purchase_ok"):
+                not_synced_prs |= line.purchase_requisition_line_ids.filtered(
+                    # Convert quantity to the same UoM as the line
+                    lambda prl: prl.requisition_id.state != "cancel"
+                    and prl.product_uom_id._compute_quantity(
+                        prl.product_qty, line.product_uom
+                    )
+                    != line.product_uom_qty
+                ).mapped("requisition_id")
+            if not_synced_prs:
+                sale.purchase_requisition_sync_warning = _(
+                    "Some lines are not synced on these Purchase Requisitions: %s",
```

Comment:
I think it is particularly interesting that at least one test should assert the warning is being displayed appropriately.